server-openid4vci/src/main/java/com/android/identity/server/openid4vci/createSession.kt

```diff
@@ -0,0 +1,71 @@
+package com.android.identity.server.openid4vci
+
+import com.android.identity.crypto.X509Cert
+import com.android.identity.flow.handler.InvalidRequestException
+import com.android.identity.flow.server.FlowEnvironment
+import com.android.identity.flow.server.Storage
+import com.android.identity.issuance.common.cache
+import com.android.identity.sdjwt.util.JsonWebKey
+import com.android.identity.util.fromBase64Url
+import jakarta.servlet.http.HttpServletRequest
+import kotlinx.io.bytestring.ByteString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonObject
+
+suspend fun createSession(environment: FlowEnvironment, req: HttpServletRequest): String {
+    // Read all parameters
+    if (req.getParameter("client_assertion_type") != ParServlet.ASSERTION_TYPE) {
+        throw InvalidRequestException("invalid parameter 'client_assertion_type'")
+    }
+    val scope = req.getParameter("scope") ?: ""
+    if (!CredentialFactory.supportedScopes.contains(scope)) {
+        throw InvalidRequestException("invalid parameter 'scope'")
+    }
+    if (req.getParameter("response_type") != "code") {
+        throw InvalidRequestException("invalid parameter 'response_type'")
+    }
+    if (req.getParameter("code_challenge_method") != "S256") {
+        throw InvalidRequestException("invalid parameter 'code_challenge_method'")
+    }
+    val redirectUri = req.getParameter("redirect_uri")
+        ?: throw InvalidRequestException("missing parameter 'redirect_uri'")
+    val clientId = req.getParameter("client_id")
+        ?: throw InvalidRequestException("missing parameter 'client_id'")
+    val codeChallenge = try {
+        ByteString(req.getParameter("code_challenge").fromBase64Url())
+    } catch (err: Exception) {
+        throw InvalidRequestException("invalid parameter 'code_challenge'")
+    }
+    val clientAssertion = req.getParameter("client_assertion")
+
+    // Check that client assertion is signed by the trusted client.
+    val sequence = clientAssertion.split("~")
+    val clientCertificate =
+        environment.cache(
+            ClientCertificate::class,
+            clientId
+        ) { configuration, resources ->
+            // So in real life this should be parameterized by clientId, as different clients will
```

Comment:
I see that this just moved from the other file, but...do we need a TODO here, to track that something needs to change someday?

Reply:
I asked the question on the clientId on mattermost, but got no answer. It seems there is no consensus on this yet. I am not sure there is something meaningful I can write there.